### PR TITLE
MDEV-8072 - encryption of Galera gcache

### PR DESCRIPTION
**Pull request by mkaruza**
* Implementation of wsrep-lib/encryption_service interface
* Serialization / de-serializaion of key

WREP startup needs to be postponed after initialization of encryption plugin. When encryption is used, key needs to be known before loading provider.

* Moved wsrep_init_startup() to be executed inside wsrep plugin, for SST methods which require to be done before SE.
* Added new plugin type MYSQL_WSREP_PLUGIN to be used for WSREP plugin, so we can initialize after encryption plugin and before SE.
* Reopen binlog index file after SST is done. Index could be opened before, but we can still receive binlog files through SST.

### CODE REVIEW

File: sql/wsrep_encryption.cc [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+ * @param input Container for serialized key
+ * @param key Pointer to encryption key
+ * @param size Size of key
+ * @param version Key version
+ */
+static void wsrep_key_serialize(std::vector<unsigned char>& input, const void* key,
+                                size_t size, unsigned int version)
+{
+  input.resize(key_version_preamble_size + size);
+  memcpy(input.data(), (void *)&version, sizeof(version));
```
Comment (janlindstrom) [inline sql/wsrep_encryption.cc]: If I correctly understand this is not be/le agnostic, but not sure if that matters.

File: sql/wsrep_encryption.cc [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+ * @param version Key version
+ * 
+ * @return 0 when key is sucessfuly de-serialized. 1 otherwise.
+ */
+static int wsrep_key_deserialize(const void* input, const size_t& input_size,
+                                 const void*& key_ptr, size_t& size,
+                                  unsigned int& version)
+{
+  bool ret= 1;
+  if(input_size >= key_version_preamble_size)
```
Comment (janlindstrom) [inline sql/wsrep_encryption.cc]: Add space after if (see also below few similar cases)

File: sql/wsrep_encryption.cc [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  {
+    memcpy(&version, input, sizeof(version));
```
Comment (janlindstrom) [inline sql/wsrep_encryption.cc]: Here also better would be some canonical form.

File: sql/wsrep_encryption.cc [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+int Wsrep_encryption_service::do_crypt(void**                ctx,
+                                       wsrep::const_buffer&  key,
+                                       const char            (*iv)[32],
+                                       wsrep::const_buffer&  input,
+                                       void*                 output,
+                                       bool                  encrypt,
+                                       bool                  last)
+{
+
+  const void* deserialized_key_ptr;
```
Comment (janlindstrom) [inline sql/wsrep_encryption.cc]: Why this can't be const unsigned char to avoid C-style cast below ?

File: sql/wsrep_encryption.cc [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    *ctx = ::malloc(encryption_ctx_size(ENCRYPTION_KEY_SYSTEM_DATA,
+                                        key_version));
+    if (*ctx == NULL)
+    {
+      throw wsrep::runtime_error("Memory not allocated in do_crypt()");
+    }
+    if(encryption_ctx_init(*ctx,
+                           (const unsigned char*)deserialized_key_ptr,
+                           deserialized_key_size,
+                           (unsigned char*)iv, MY_AES_BLOCK_SIZE,
```
Comment (janlindstrom) [inline sql/wsrep_encryption.cc]: Cant we use unsigned char also on function parameters to avoid this C-style cast also ?

**Comment by svoj** [comment]
@mkaruza I don't completely get why did you have to add new plugin type. How is it different from deamon plugin type?

**Comment by janlindstrom** [comment]
I do not like the name MYSQL_WSREP_PLUGIN and can't you use some of the existing WSREP plugins like WSREP_INFO ?

File: sql/wsrep_encryption.cc (hunk 2 above)
Comment (izenish) [reply sql/wsrep_encryption.cc]: Does it make a difference?

Comment (mkaruza) [reply sql/wsrep_encryption.cc]: Added space

File: sql/wsrep_encryption.cc (hunk 3 above)
Comment (mkaruza) [reply sql/wsrep_encryption.cc]: Added space

File: sql/wsrep_encryption.cc (hunk 4 above)
Comment (mkaruza) [reply sql/wsrep_encryption.cc]: It can be, changed

**Comment by mkaruza** [comment]
@svoj @janlindstrom Ordering of plugin initialization matters and we need to initialize wsrep/galera after ENCRYPTION and before SE plugin initialization - so patch creates new plugin type that is initialized in  between.

**Comment by janlindstrom** [comment]
Can you change e.g. WSREP_INFO Plugin to be initialized after encryption and before SE and add that encryption stuff ? If not then WSREP_ENCRYPTION would make more sense.

**Comment by mkaruza** [comment]
@janlindstrom WSREP_INFO plugin is type MYSQL_INFORMATION_SCHEMA_PLUGIN, which is initialized after MYSQL_STORAGE_ENGINE_PLUGIN. There are other plugins of this type in code base.
WSREP_ENCRYPTION could be candidate name, but plugin is used only to postpone call to `wsrep_init_startup`.

**Comment by janlindstrom** [comment]
Is WSREP_PLUGIN fine ?

**Comment by mkaruza** [comment]
It is already named WSREP_PLUGIN

**Comment by svoj** [comment]
I need to think about adding new plugin time. I don't like this approach, but unfortunately have nothing to suggest at the moment. May be @vuvova could comment on this?

File: sql/mysqld.cc [hunk 6]
```diff
@@ -5110,8 +5110,7 @@ static int init_server_components()
     if (opt_bootstrap) // bootsrap option given - disable wsrep functionality
     {
       wsrep_provider_init(WSREP_NONE);
-      if (wsrep_init())
-        unireg_abort(1);
+      if (wsrep_init()) unireg_abort(1);
```
Comment (svoj) [inline sql/mysqld.cc]: Please revert this change, it doesn't follow MariaDB coding style.

File: sql/mysqld.cc [hunk 7]
```diff
@@ -5206,6 +5214,15 @@ static int init_server_components()
...
   plugins_are_initialized= TRUE;  /* Don't separate from init function */
 
+#ifdef WITH_WSREP
+  /* In case of SST before SE initialization is done from wsrep plugin,
+     if error occured we are aborting */
+  if (wsrep_startup_state == WSREP_STARTUP_STATE_MUST_ABORT)
+  {
+    unireg_abort(1);
+  }
+#endif
```
Comment (svoj) [inline sql/mysqld.cc]: Why this block has to be here? Why not in `wsrep_plugin_init()`, like:
```
if (wsrep_init_startup(true))
  return wsrep_startup_state == WSREP_STARTUP_STATE_MUST_ABORT;
```

File: sql/sql_plugin.cc [hunk 8]
```diff
@@ -91,6 +91,9 @@ const LEX_CSTRING plugin_type_names[MYSQL_MAX_PLUGIN_TYPE_NUM]=
   { STRING_WITH_LEN("AUTHENTICATION") },
   { STRING_WITH_LEN("PASSWORD VALIDATION") },
   { STRING_WITH_LEN("ENCRYPTION") }
+#ifdef WITH_WSREP
+  , { STRING_WITH_LEN("WSREP") }
+#endif
```
Comment (svoj) [inline sql/sql_plugin.cc]: If we decide to stay with the new plugin type approach, it should not be under ifdefs. We can't reuse this plugin number for anything else anyway.

**Comment by janlindstrom** [comment]
Merged.
